app/src/main/java/com/github/drawyourpath/bootcamp/authentication/FirebaseAuth.kt

```diff
@@ -0,0 +1,236 @@
+            if (user == null) {
+                return null
+            }
+            return object : User {
+                override fun getDisplayName(): String {
+                    return user.displayName ?: ""
+                }
+
+                override fun getEmail(): String {
+                    return user.email ?: ""
+                }
+
+                override fun getPhoneNumber(): String {
+                    return user.phoneNumber ?: ""
+                }
+
+                override fun getPhotoUrl(): Uri? {
+                    return user.photoUrl
+                }
+
+                override fun getUid(): String {
+                    return user.uid
+                }
+
+                override fun isAnonymous(): Boolean {
+                    return user.isAnonymous
+                }
+            }
+        }
+
+        fun getUser(): User? {
+            return convertUser(FirebaseAuth.getInstance().currentUser)
+        }
+    }
+
+    private var currCallback: AuthCallback? = null
+
+    private fun setCurrCallback(callback: AuthCallback): Boolean
+    {
+        if (currCallback != null) {
+            callback(null, Exception("An operation is still pending."))
+            return false
+        }
+        currCallback = callback
+        return true
+    }
+
+    private fun consumeCurrCallback(user: User?, error: Exception?) {
+        if (currCallback != null) {
+            currCallback!!(user, error)
+            currCallback = null
+        }
+    }
+
+    override fun getUser(): User? {
+        return convertUser(FirebaseAuth.getInstance().currentUser)
+    }
+
+    override fun loginWithGoogle(activity: Activity, callback: AuthCallback) {
+        if (!setCurrCallback(callback)) {
+            return
+        }
+
+        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
+            .requestIdToken(activity.getString(R.string.server_client_id))
+            .requestEmail()
+            .build()
+        val client = GoogleSignIn.getClient(activity, gso)
+
+        client.signInIntent.also {
+            activity.startActivityForResult(it, REQ_GSI)
+        }
+    }
+
+    override fun loginWithEmail(email: String, password: String, callback: AuthCallback) {
+        TODO("Not yet implemented")
+    }
+
+    override fun loginAnonymously(callback: AuthCallback) {
+        TODO("Not yet implemented")
+    }
+
+    override fun registerWithEmail(email: String, password: String, callback: AuthCallback) {
+        auth
+            .createUserWithEmailAndPassword(email, password)
+            .addOnSuccessListener {res ->
+                callback(convertUser(res.user!!), null)
+            }
+            .addOnFailureListener { exc ->
+                callback(null, exc)
+            }
+    }
+
+    override fun registerWithGoogle(activity: Activity, callback: AuthCallback) {
+        loginWithGoogle(activity, callback)
+    }
```

Comment:
Why is registerWithGoogle the same as loginWithGoogle?

Reply:
The Google Sign-In SDK doesn't make a difference between registering and signing in. Registering is just the first sign in.